core/src/main/scala/cats/collections/Predicate.scala

```diff
@@ -1,17 +1,29 @@
 package cats.collections
 
 import cats._
+import cats.data.AndThen
 
 /**
  * An intensional set, which is a set which instead of enumerating its
  * elements as a extensional set does, it is defined by a predicate
  * which is a test for membership.
+ *
+ * All operations in this class are implemented in a stack-safe way.
  */
-abstract class Predicate[-A] extends scala.Function1[A, Boolean] { self =>
+final class Predicate[-A] private(private val p: AndThen[A, Eval[Boolean]]) { self =>
```

Comment:
I'm still pretty skeptical of the need for both AndThen and Eval. Eval comes with a big performance penalty, like 100x isn't uncommon.

Is blowing the stack very common here?

Next, Why not just make `type Predicate[-A] = Kleisli[Eval, A, Boolean]` and then have some methods on that if you want to have a stack safe predicate? Why a new class?

Reply:
Regarding performance, I'm not seeing any 100x: I posted a benchmark in the opening post and it seems to be at 2x worst case scenario. I might be missing something though.

Regarding the need for `Eval`, since we are dealing with tree-shaped expressions, there is a need for some type of trampoline like `Eval` if stack safety is to be achieved. (A `Defer[_]` to be precise) Maybe `scala.util.control.TailCalls.TailRec` would be preferable since we dont need all the laziness modes?

Regarding `AndThen`, it makes contramap stack-safe, and it's a thin layer over functions that doesn't even kick in unless a lot of composition is done, so I think it's actually more performant in this role than Kleisli because Kleisli leans on the trampoline. I'll come with benchmarks though.

Regarding whether it's common to blow the stack here, I come at it from the perspective that it shouldn't be possible at all in a Cats library. It should be as efficient as possible while remaining SOE-less. That's my interpretation of the "Efficency" motivation on the [Cats page](https://typelevel.org/cats/motivations.html#efficiency).

Regarding the class question that's a good point. I tried turning the class into an opaque newtype a couple of months ago, but it wouldn't compile under 2.11. Now that 2.11 support has been dropped we could go that route, or do a transparent type alias.

Reply:
Regarding trampolines, I realized the most suitable trampoline is probably a dedicated one: we only use Eval.{True, False}, defer, and flatMap, so rolling my own should be fairly easy and would have built-in unboxed booleans. We shall see on the benchmarks.

Reply:
Sorry, I overlooked the benchmarks, those look good to me. My concerns are removed by those.

I don't think a custom trampoline can speed this up (maybe I'm wrong) since there are only two Booleans boxing them is almost free, and `Eval.True` and `Eval.False` have already boxed them. Eval is faster than scala TailCall and cats.free.Free (by a significant margin) so beating it seems tough to me.

Reply:
I'm now seeing a combination of the trampoline and the AST, similar to a trampoline with a monadReader capability. It's probably a tiny bit faster, it's more elegant and allows to apply laws so I'll try that next. Thanks you for stimulating my thought process.